File: apps/viewer/lib/admin_client.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'config.dart';

/// Calls the broker's admin/ops endpoints. Every request carries the signed-in user's
/// Supabase access token as a Bearer; the broker verifies it belongs to an app_admin
/// (RLS-independent — the broker checks app_admins with the service-role key).
class AdminException implements Exception {
  final String message;
  AdminException(this.message);
  @override
  String toString() => message;
}

class AdminClient {
  AdminClient(this.accessToken);
  final String accessToken;

  bool get available => brokerUrl.isNotEmpty;

  Map<String, String> get _headers => {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $accessToken',
      };

  Future<Map<String, dynamic>> _get(String path) async {
    _ensure();
    final r = await http
        .get(Uri.parse('$brokerUrl$path'), headers: _headers)
        .timeout(const Duration(seconds: 40));
    return _decode(r);
  }

  Future<Map<String, dynamic>> _post(String path, [Map<String, dynamic>? body]) async {
    _ensure();
    final r = await http
        .post(Uri.parse('$brokerUrl$path'),
            headers: _headers, body: jsonEncode(body ?? const {}))
        .timeout(const Duration(seconds: 40));
    return _decode(r);
  }

  void _ensure() {
    if (!available) throw AdminException('Ops console needs BROKER_URL configured.');
  }

  Map<String, dynamic> _decode(http.Response r) {
    Map<String, dynamic> data;
    try {
      data = jsonDecode(r.body) as Map<String, dynamic>;
    } catch (_) {
      throw AdminException('Broker error (${r.statusCode}).');
    }
    if (r.statusCode == 403) throw AdminException('Not authorized (admins only).');
    if (r.statusCode >= 400) {
      throw AdminException((data['detail'] ?? 'Request failed (${r.statusCode}).').toString());
    }
    return data;
  }

  Future<Map<String, dynamic>> summary() => _get('/admin/summary');
  Future<Map<String, dynamic>> actions() => _get('/admin/actions');
  Future<Map<String, dynamic>> diagnostics() => _get('/admin/diagnostics');
  Future<Map<String, dynamic>> run(String workflow, {Map<String, dynamic>? inputs}) =>
      _post('/admin/actions/run', {'workflow': workflow, if (inputs != null) 'inputs': inputs});
  Future<Map<String, dynamic>> rerun(int runId) => _post('/admin/actions/$runId/rerun');
  Future<Map<String, dynamic>> invite(String email) => _post('/admin/invite', {'email': email});
}

File: native/android/app/src/main/java/org/membercovenantpath/viewer/data/AdminClient.kt
package org.membercovenantpath.viewer.data

import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject

/** Mirrors admin_client.dart's AdminException. */
class AdminException(message: String) : Exception(message)

/**
 * Calls the broker's admin/ops endpoints. Every request carries the signed-in user's Supabase access
 * token; the broker verifies it belongs to an app_admin (service-role check, RLS-independent).
 * Ported 1:1 from admin_client.dart.
 */
class AdminClient(private val accessToken: String) {

    val available: Boolean get() = AppConfig.brokerAvailable
    private val base: String get() = AppConfig.brokerUrl

    private fun ensure() {
        if (!available) throw AdminException("Ops console needs BROKER_URL configured.")
    }

    private suspend fun get(path: String): JsonObject {
        ensure()
        val resp = Net.getJson("$base$path", accessToken)
        return decode(resp.status.value, Net.bodyObject(resp))
    }

    private suspend fun post(path: String, body: JsonObject = JsonObject(emptyMap())): JsonObject {
        ensure()
        val resp = Net.postJson("$base$path", body, accessToken)
        return decode(resp.status.value, Net.bodyObject(resp))
    }

    private fun decode(status: Int, data: JsonObject): JsonObject {
        if (status == 403) throw AdminException("Not authorized (admins only).")
        if (status >= 400) throw AdminException(data.str("detail") ?: "Request failed ($status).")
        return data
    }

    suspend fun summary() = get("/admin/summary")
    suspend fun actions() = get("/admin/actions")
    suspend fun diagnostics() = get("/admin/diagnostics")
    suspend fun enrolledStakes() = get("/admin/enrolled-stakes")
    suspend fun revokeStake(stakeId: String) = post("/admin/stakes/$stakeId/revoke")
    /** Wipe a stake's member data (keeps the stake + roles + credential; repopulates next sync). */
    suspend fun wipeStakeData(stakeId: String) = post("/admin/stakes/$stakeId/wipe-data")
    /** Remove a stake completely (credential + members + roles + the stake row). Irreversible. */
    suspend fun removeStake(stakeId: String) = post("/admin/stakes/$stakeId/remove")
    suspend fun run(workflow: String, inputs: JsonObject? = null) =
        post("/admin/actions/run", buildJsonObject {
            put("workflow", JsonPrimitive(workflow))
            if (inputs != null) put("inputs", inputs)
        })
    suspend fun rerun(runId: Long) = post("/admin/actions/$runId/rerun")
    suspend fun invite(email: String) =
        post("/admin/invite", buildJsonObject { put("email", JsonPrimitive(email)) })
    suspend fun feedback(title: String, body: String) =
        post("/feedback", buildJsonObject {
            put("title", JsonPrimitive(title)); put("body", JsonPrimitive(body))
        })
}
